sale_coupon_advanced/models/sale_order.py

```diff
@@ -1,11 +1,67 @@
 # Copyright 2020 Camptocamp SA
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
-from odoo import models
+
+from odoo import _, models
 
 
 class SaleOrder(models.Model):
     _inherit = "sale.order"
 
+    def _get_reward_line_values(self, program):
+        """ Add the rewarded product if a reward line has been found for this
+            product
+        """
+        reward_values = super()._get_reward_line_values(program)
+        reward_product = program.reward_product_id
+        discount_product = program.discount_line_product_id
+        for product in reward_values:
+            product_id = product["product_id"]
+            is_reward = product["is_reward_line"]
+            if product_id == discount_product.id and is_reward:
+                taxes = reward_product.taxes_id
+                if self.fiscal_position_id:
+                    taxes = self.fiscal_position_id.map_tax(taxes)
+                reward_values += [
+                    {
+                        "sequence": 998,  # Put this line before discount
+                        "name": reward_product.name,
+                        "product_id": reward_product.id,
+                        "price_unit": reward_product.lst_price,
+                        "is_reward_line": False,
+                        "product_uom_qty": program.reward_product_quantity,
+                        "product_uom": reward_product.uom_id.id,
+                        "tax_id": [(4, tax.id, False) for tax in taxes],
+                    }
+                ]
+        return reward_values
+
+    def _get_reward_values_product(self, program):
+        """ Send a free product when
+            `sale_coupon_program.is_reward_product_forced` is selected
+            and reward product not already ordered
+            (added later in _get_reward_line_values)
+        """
+        if not program.is_reward_product_forced and not self._is_reward_in_order_lines(
+            program
+        ):
```

Comment:
```suggestion
        if (program.is_reward_product_forced or self._is_reward_in_order_lines(program):
```
I like to store those conditions in a very explicitly named variable.
It makes the code easier to read and makes it easier to write it on one line.

```suggestion
        already_applied_program = self._is_reward_in_order_lines(program)
        if (program.is_reward_product_forced or already_applied_program):
```

Reply:
It's quite a personal programming aspect i think. But its interesting to avoid two lines in this case actually ^^

Reply:
It is more interesting to make the code more understandable.
Indentation is a really nice way to make the code more understandable.
Having "):" on a single line right before an indented block is not very helpful.
But yes, if black says it's okay, that should be fine.
But IMO, if black splits a condition during pre-commit, you should create a few more variables.

Reply:
Plus, your boolean expression is simplifiable here :
`not bool and not bool` is equal to `bool or bool`

Reply:
Mme Mequignon, vous chipotez !

Reply:
Fixed